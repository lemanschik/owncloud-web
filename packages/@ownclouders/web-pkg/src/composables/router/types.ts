import { RouteMeta } from 'vue-router'

type Dictionary<T> = { [key: string]: T }

export type QueryValue = string | (string | null)[]
export type LocationQuery = Dictionary<QueryValue>

export type ParamValue = string

export const authContextValues = ['anonymous', 'user', 'idp', 'publicLink', 'hybrid'] as const
export type AuthContext = typeof authContextValues[number]

export interface WebRouteMeta extends RouteMeta {
  title?: string
  authContext?: AuthContext
  patchCleanPath?: boolean
  contextQueryItems?: string[]
}
