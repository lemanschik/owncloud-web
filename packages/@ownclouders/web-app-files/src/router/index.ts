import { RouteLocationNamedRaw } from 'vue-router'

import {
  buildRoutes as buildCommonRoutes,
  isLocationCommonActive,
  createLocationCommon
} from './common'
import { buildRoutes as buildDeprecatedRoutes, isLocationActive } from './deprecated'
import {
  buildRoutes as buildPublicRoutes,
  createLocationPublic,
  isLocationPublicActive
} from './public'
import { RouteComponents } from './router'
import {
  buildRoutes as buildSharesRoutes,
  isLocationSharesActive,
  createLocationShares
} from './shares'
import {
  buildRoutes as buildSpacesRoutes,
  isLocationSpacesActive,
  createLocationSpaces
} from './spaces'
import {
  buildRoutes as buildTrashRoutes,
  isLocationTrashActive,
  createLocationTrash
} from './trash'
import type { ActiveRouteDirectorFunc } from './utils'

const ROOT_ROUTE = {
  name: 'root',
  path: '/',
  redirect: (to) => createLocationSpaces('files-spaces-generic', to)
}

const buildRoutes = (components: RouteComponents): RouteLocationNamedRaw[] => [
  ROOT_ROUTE,
  ...buildCommonRoutes(components),
  ...buildSharesRoutes(components),
  ...buildPublicRoutes(components),
  ...buildSpacesRoutes(components),
  ...buildTrashRoutes(components),
  ...buildDeprecatedRoutes()
]

export {
  createLocationCommon,
  createLocationShares,
  createLocationSpaces,
  createLocationPublic,
  isLocationCommonActive,
  isLocationSharesActive,
  isLocationSpacesActive,
  isLocationPublicActive,
  isLocationActive,
  isLocationTrashActive,
  createLocationTrash,
  buildRoutes,
  ActiveRouteDirectorFunc
}
