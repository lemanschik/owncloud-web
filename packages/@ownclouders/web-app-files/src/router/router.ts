import { defineComponent } from 'vue'

/**
 * we need to inject the vue files into the route builders,
 * this is because we also import the provided helpers from other js|ts files
 * like mixins, rollup seems to have a problem to import files which contain vue file imports
 * into js files which then again get imported by other vue files...
 */

type Component = ReturnType<typeof defineComponent>

export interface RouteComponents {
  App: Component
  Favorites: Component
  FilesDrop: Component
  SearchResults: Component
  Shares: {
    SharedWithMe: Component
    SharedWithOthers: Component
    SharedViaLink: Component
  }
  Spaces: {
    DriveResolver: Component
    Projects: Component
  }
  Trash: {
    Overview: Component
  }
}
