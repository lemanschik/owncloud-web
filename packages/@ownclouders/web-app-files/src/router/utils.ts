import { Router, RouteLocationNamedRaw } from 'vue-router'
import merge from 'lodash-es/merge'
import { unref } from 'vue'

export interface ActiveRouteDirectorFunc<T extends string> {
  (router: Router, ...comparatives: T[]): boolean
}

/**
 * helper function to find out if comparative route location is active or not.
 * it uses vue router resolve to do so.
 *
 * @param router
 * @param comparatives
 */
export const isLocationActive = (
  router: Router,
  ...comparatives: [RouteLocationNamedRaw, ...RouteLocationNamedRaw[]]
): boolean => {
  // FIXME: router.resolve cleans the path. we don't need it, if we can rely on
  // router.currentRoute to not have slashs encoded for paths
  const { href: currentHref } = router.resolve(unref(router.currentRoute))
  return comparatives
    .map((comparative) => {
      const { href: comparativeHref } = router.resolve({
        ...comparative
        // ...(comparative.name && { name: resolveRouteName(comparative.name) })
      })

      /**
       * Href might be '/' or '#/' if router is not able to resolve the proper path.
       * This happens if the we don't pass a param which is defined in the route configuration, for example:
       * path: user/:id
       *
       * This implies that the comparative route is not active
       **/
      if (comparativeHref === '/' || comparativeHref === '#/') {
        return false
      }
      return currentHref.startsWith(comparativeHref)
    })
    .some(Boolean)
}

/**
 * wraps isLocationActive to be used as a closure,
 * the resulting closure then can be used to check a location against the defined set of director locations
 *
 * @param defaultComparatives
 */
export const isLocationActiveDirector = <T extends string>(
  ...defaultComparatives: [RouteLocationNamedRaw, ...RouteLocationNamedRaw[]]
): ActiveRouteDirectorFunc<T> => {
  return (router: Router, ...comparatives: T[]): boolean => {
    if (!comparatives.length) {
      return isLocationActive(router, ...defaultComparatives)
    }

    const [first, ...rest] = comparatives.map((name) => {
      const match = defaultComparatives.find((c) => c.name === name)

      if (!match) {
        throw new Error(`unknown comparative '${name}'`)
      }

      return match
    })

    return isLocationActive(router, first, ...rest)
  }
}

/**
 * just a dummy function to trick gettext tools
 *
 * @param msg
 */
export function $gettext(msg: string): string {
  return msg
}

/**
 * create a location with attached default values
 *
 * @param name
 * @param locations
 */
export const createLocation = (
  name: string,
  ...locations: RouteLocationNamedRaw[]
): RouteLocationNamedRaw =>
  merge(
    {},
    {
      name
    },
    ...locations.map((location) => ({
      ...(location.params && { params: location.params }),
      ...(location.query && { query: location.query })
    }))
  )
