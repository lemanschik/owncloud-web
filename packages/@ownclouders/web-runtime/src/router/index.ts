import qs from 'qs'
import AccessDeniedPage from '../pages/accessDenied.vue'
import Account from '../pages/account.vue'
import LoginPage from '../pages/login.vue'
import LogoutPage from '../pages/logout.vue'
import OidcCallbackPage from '../pages/oidcCallback.vue'
import ResolvePublicLinkPage from '../pages/resolvePublicLink.vue'
import ResolvePrivateLinkPage from '../pages/resolvePrivateLink.vue'
import { setupAuthGuard } from './setupAuthGuard'
import { patchRouter } from './patchCleanPath'
import { createWebHashHistory, createWebHistory, createRouter } from 'vue-router'

export * from './helpers'
export { createRouter } from 'vue-router'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

export const base = document.querySelector('base')
export const router = patchRouter(
  createRouter({
    parseQuery(query) {
      return qs.parse(query, {
        allowDots: true
      })
    },
    stringifyQuery(obj) {
      return qs.stringify(obj, {
        allowDots: true
      })
    },
    history: (base && createWebHistory(new URL(base.href).pathname)) || createWebHashHistory(),
    routes: [
      {
        path: '/login',
        name: 'login',
        component: LoginPage,
        meta: { title: $gettext('Login'), authContext: 'anonymous' }
      },
      {
        path: '/logout',
        name: 'logout',
        component: LogoutPage,
        meta: { title: $gettext('Logout'), authContext: 'anonymous' }
      },
      {
        path: '/oidc-callback',
        name: 'oidcCallback',
        component: OidcCallbackPage,
        meta: { title: $gettext('Oidc callback'), authContext: 'anonymous' }
      },
      {
        path: '/oidc-silent-redirect',
        name: 'oidcSilentRedirect',
        component: OidcCallbackPage,
        meta: { title: $gettext('Oidc redirect'), authContext: 'anonymous' }
      },
      {
        path: '/f/:fileId',
        name: 'resolvePrivateLink',
        component: ResolvePrivateLinkPage,
        meta: { title: $gettext('Private link'), authContext: 'user' }
      },
      {
        path: '/s/:token',
        name: 'resolvePublicLink',
        component: ResolvePublicLinkPage,
        meta: { title: $gettext('Public link'), authContext: 'anonymous' }
      },
      {
        path: '/access-denied',
        name: 'accessDenied',
        component: AccessDeniedPage,
        meta: { title: $gettext('Access denied'), authContext: 'anonymous' }
      },
      {
        path: '/account',
        name: 'account',
        component: Account,
        meta: { title: $gettext('Account'), authContext: 'user' }
      }
    ]
  })
)

setupAuthGuard(router)
