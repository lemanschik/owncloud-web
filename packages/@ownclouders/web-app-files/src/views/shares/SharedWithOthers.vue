<template>
  <div class="oc-flex">
    <files-view-wrapper>
      <app-bar :has-shares-navigation="true" :side-bar-open="sideBarOpen" />
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <no-content-message
          v-if="isEmpty"
          id="files-shared-with-others-empty"
          class="files-empty"
          icon="reply"
        >
          <template #message>
            <span v-translate> You have not shared any resources with other people. </span>
          </template>
        </no-content-message>
        <resource-table
          v-else
          id="files-shared-with-others-table"
          v-model:selectedIds="selectedResourcesIds"
          class="files-table"
          :class="{ 'files-table-squashed': sideBarOpen }"
          :fields-displayed="['name', 'sharedWith', 'sdate']"
          :are-thumbnails-displayed="displayThumbnails"
          :are-paths-displayed="true"
          :resources="paginatedResources"
          :header-position="fileListHeaderY"
          :sort-by="sortBy"
          :sort-dir="sortDir"
          @file-click="triggerDefaultAction"
          @row-mounted="rowMounted"
          @sort="handleSort"
        >
          <template #contextMenu="{ resource }">
            <context-actions
              v-if="isResourceInSelection(resource)"
              :action-options="{ space: getSpace(resource), resources: selectedResources }"
            />
          </template>
          <template #footer>
            <pagination :pages="paginationPages" :current-page="paginationPage" />
            <list-info
              v-if="paginatedResources.length > 0"
              class="oc-width-1-1 oc-my-s"
              :files="totalFilesCount.files"
              :folders="totalFilesCount.folders"
            />
          </template>
        </resource-table>
      </template>
    </files-view-wrapper>
    <side-bar
      :open="sideBarOpen"
      :active-panel="sideBarActivePanel"
      :space="selectedResourceSpace"
    />
  </div>
</template>

<script lang="ts">
import { mapGetters, mapState, mapActions } from 'vuex'

import { useFileActions } from '../../composables/actions/files/useFileActions'
import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageDimension, ImageType } from 'web-pkg/src/constants'
import { debounce } from 'lodash-es'

import ResourceTable from '../../components/FilesList/ResourceTable.vue'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppBar from '../../components/AppBar/AppBar.vue'
import ListInfo from '../../components/FilesList/ListInfo.vue'
import Pagination from '../../components/FilesList/Pagination.vue'
import ContextActions from '../../components/FilesList/ContextActions.vue'
import SideBar from '../../components/SideBar/SideBar.vue'
import FilesViewWrapper from '../../components/FilesViewWrapper.vue'

import { useResourcesViewDefaults } from '../../composables'
import { defineComponent } from 'vue'
import { Resource } from 'web-client'
import { SpaceResource } from 'web-client/src/helpers'
import { useStore } from 'web-pkg/src/composables'
import { getSpaceFromResource } from 'web-app-files/src/helpers/resource/getSpace'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    FilesViewWrapper,
    AppBar,
    ResourceTable,
    AppLoadingSpinner,
    NoContentMessage,
    ListInfo,
    Pagination,
    ContextActions,
    SideBar
  },

  setup() {
    const store = useStore()
    const getSpace = (resource: Resource): SpaceResource => {
      return getSpaceFromResource({ spaces: store.getters['runtime/spaces/spaces'], resource })
    }

    return {
      ...useFileActions(),
      ...useResourcesViewDefaults<Resource, any, any[]>(),
      getSpace
    }
  },

  computed: {
    ...mapState(['app']),
    ...mapState('Files', ['files']),
    ...mapGetters('Files', ['totalFilesCount']),
    ...mapGetters(['configuration', 'user']),

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    displayThumbnails() {
      return !this.configuration?.options?.disablePreviews
    }
  },

  async created() {
    await this.loadResourcesTask.perform()
    this.scrollToResourceFromRoute(this.paginatedResources)
  },

  beforeUnmount() {
    visibilityObserver.disconnect()
  },

  methods: {
    ...mapActions('Files', ['loadPreview', 'loadAvatars']),

    rowMounted(resource, component) {
      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadAvatars({ resource, clientService: this.$clientService })

        if (!this.displayThumbnails) {
          return
        }

        this.loadPreview({
          previewService: this.$previewService,
          space: this.getSpace(resource),
          resource,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
