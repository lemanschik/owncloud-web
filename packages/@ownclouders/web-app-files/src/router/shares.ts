import { RouteComponents } from './router'
import { RouteLocationNamedRaw, RouteRecordRaw } from 'vue-router'
import { createLocation, isLocationActiveDirector, $gettext } from './utils'

type shareTypes =
  | 'files-shares'
  | 'files-shares-with-me'
  | 'files-shares-with-others'
  | 'files-shares-via-link'

export const createLocationShares = (name: shareTypes, location = {}): RouteLocationNamedRaw =>
  createLocation(name, location)

export const locationShares = createLocationShares('files-shares')
export const locationSharesWithMe = createLocationShares('files-shares-with-me')
export const locationSharesWithOthers = createLocationShares('files-shares-with-others')
export const locationSharesViaLink = createLocationShares('files-shares-via-link')

export const isLocationSharesActive = isLocationActiveDirector<shareTypes>(
  locationSharesWithMe,
  locationSharesWithOthers,
  locationSharesViaLink
)

export const buildRoutes = (components: RouteComponents): RouteRecordRaw[] => [
  {
    name: locationShares.name,
    path: '/shares',
    component: components.App,
    redirect: locationSharesWithMe,
    children: [
      {
        name: locationSharesWithMe.name,
        path: 'with-me',
        component: components.Shares.SharedWithMe,
        meta: {
          authContext: 'user',
          title: $gettext('Files shared with me')
        }
      },
      {
        name: locationSharesWithOthers.name,
        path: 'with-others',
        component: components.Shares.SharedWithOthers,
        meta: {
          authContext: 'user',
          title: $gettext('Files shared with others')
        }
      },
      {
        name: locationSharesViaLink.name,
        path: 'via-link',
        component: components.Shares.SharedViaLink,
        meta: {
          authContext: 'user',
          title: $gettext('Files shared via link')
        }
      }
    ]
  }
]
