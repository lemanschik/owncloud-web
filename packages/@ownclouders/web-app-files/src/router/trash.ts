import { RouteComponents } from './router'
import { RouteLocationNamedRaw, RouteRecordRaw } from 'vue-router'
import { $gettext, createLocation, isLocationActiveDirector } from './utils'

type trashTypes = 'files-trash-generic' | 'files-trash-overview'

export const createLocationTrash = (name: trashTypes, location = {}): RouteLocationNamedRaw =>
  createLocation(name, location)

export const locationTrashGeneric = createLocationTrash('files-trash-generic')

export const locationTrashOverview = createLocationTrash('files-trash-overview')

export const isLocationTrashActive = isLocationActiveDirector<trashTypes>(
  locationTrashGeneric,
  locationTrashOverview
)

export const buildRoutes = (components: RouteComponents): RouteRecordRaw[] => [
  {
    path: '/trash',
    component: components.App,
    children: [
      {
        path: 'overview',
        name: locationTrashOverview.name,
        component: components.Trash.Overview,
        meta: {
          authContext: 'user',
          title: $gettext('Trash overview')
        }
      },
      {
        name: locationTrashGeneric.name,
        path: ':driveAliasAndItem(.*)?',
        component: components.Spaces.DriveResolver,
        meta: {
          authContext: 'user',
          patchCleanPath: true
        }
      }
    ]
  }
]
