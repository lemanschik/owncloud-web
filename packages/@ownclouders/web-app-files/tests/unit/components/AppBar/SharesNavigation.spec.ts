import SharesNavigation from '../../../../src/components/AppBar/SharesNavigation.vue'
import { locationSharesWithMe } from 'web-app-files/src/router/shares'
import { mock } from 'jest-mock-extended'
import { RouteRecordNormalized } from 'vue-router'
import {
  createStore,
  defaultPlugins,
  defaultStubs,
  shallowMount,
  defaultStoreMockOptions,
  defaultComponentMocks,
  RouteLocation
} from 'web-test-helpers'

const routes = [
  mock<RouteRecordNormalized>({
    path: '/files/shares/with-me/',
    name: 'files-shares-with-me'
  }),
  mock<RouteRecordNormalized>({
    path: '/files/shares/with-others/',
    name: 'files-shares-with-others'
  }),
  mock<RouteRecordNormalized>({
    path: '/files/shares/via-link/',
    name: 'files-shares-via-link'
  })
]

describe('SharesNavigation component', () => {
  it('renders a shares navigation for both mobile and a desktop viewports', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper({ currentRouteName = locationSharesWithMe.name } = {}) {
  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)
  const mocks = defaultComponentMocks({
    currentRoute: mock<RouteLocation>({ name: currentRouteName })
  })
  mocks.$router.getRoutes.mockImplementation(() => routes)
  return {
    storeOptions,
    mocks,
    wrapper: shallowMount(SharesNavigation, {
      global: {
        stubs: defaultStubs,
        renderStubDefaultSlot: true,
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
