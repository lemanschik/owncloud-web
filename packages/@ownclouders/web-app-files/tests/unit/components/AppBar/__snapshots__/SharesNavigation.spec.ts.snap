// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharesNavigation component renders a shares navigation for both mobile and a desktop viewports 1`] = `
<nav aria-label="Shares pages navigation" class="oc-py-s" id="shares-navigation">
  <oc-list-stub class="oc-flex oc-visible@s" raw="false">
    <li>
      <oc-button-stub appearance="raw" class="oc-mr-m oc-py-s shares-nav-desktop" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" to="/files/shares/with-me/" type="router-link" variation="passive">
        <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="share-forward" size="small" type="span" variation="passive"></oc-icon-stub>
        <span>Shared with me</span>
      </oc-button-stub>
    </li>
    <li>
      <oc-button-stub appearance="raw" class="oc-mr-m oc-py-s shares-nav-desktop" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" to="/files/shares/with-others/" type="router-link" variation="passive">
        <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="reply" size="small" type="span" variation="passive"></oc-icon-stub>
        <span>Shared with others</span>
      </oc-button-stub>
    </li>
    <li>
      <oc-button-stub appearance="raw" class="oc-mr-m oc-py-s shares-nav-desktop" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" to="/files/shares/via-link/" type="router-link" variation="passive">
        <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="link" size="small" type="span" variation="passive"></oc-icon-stub>
        <span>Shared via link</span>
      </oc-button-stub>
    </li>
  </oc-list-stub>
  <div class="oc-hidden@s">
    <oc-button-stub appearance="raw" disabled="false" gapsize="medium" id="shares_navigation_mobile" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
      <span>Shared with me</span>
      <oc-icon-stub accessiblelabel="" color="" filltype="line" name="arrow-down-s" size="small" type="span" variation="passive"></oc-icon-stub>
    </oc-button-stub>
    <oc-drop-stub closeonclick="true" dropid="oc-drop-1" isnested="false" mode="click" paddingsize="small" position="bottom-start" toggle="#shares_navigation_mobile">
      <oc-list-stub raw="false">
        <li>
          <oc-button-stub appearance="raw-inverse" class="oc-my-xs shares-nav-mobile oc-background-primary-gradient" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" to="/files/shares/with-me/" type="router-link" variation="primary">
            <span class="icon-box icon-box-active">
              <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="share-forward" size="medium" type="span" variation="passive"></oc-icon-stub>
            </span>
            <span>Shared with me</span>
          </oc-button-stub>
        </li>
        <li>
          <oc-button-stub appearance="raw" class="oc-my-xs shares-nav-mobile" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" to="/files/shares/with-others/" type="router-link" variation="passive">
            <span class="icon-box">
              <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="reply" size="medium" type="span" variation="passive"></oc-icon-stub>
            </span>
            <span>Shared with others</span>
          </oc-button-stub>
        </li>
        <li>
          <oc-button-stub appearance="raw" class="oc-my-xs shares-nav-mobile" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" to="/files/shares/via-link/" type="router-link" variation="passive">
            <span class="icon-box">
              <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="link" size="medium" type="span" variation="passive"></oc-icon-stub>
            </span>
            <span>Shared via link</span>
          </oc-button-stub>
        </li>
      </oc-list-stub>
    </oc-drop-stub>
  </div>
</nav>
`;
